# What is Assessing Data?

##   The Data analysis process consists of following steps:-
    1. Asking Questions
    2. Data Wrangling
    
       a. Gathering Data
          
            i.    CSV Files
            ii.   APIs
            iii.  Web Scraping
            iv.   Databases
       b. Assessing Data
       c. Cleaning Data
    
    3. Exploratory Data Analysis
    4. Drawing Conclusion
    5. Communicating Results

# Types of Unclean Data

## There are 2 kind of unclean Data

1. Dirty Data(Data with Quality issues):
    Dirty data, also known as low quality data. Low quality data has content issues

2. Messy Data( Data with tidiness issues):
    Messy Data, also known as untidy data. Untidy data has structural issues.

# 1.Dirty Data

## When the Data has following of isssues:-
    1. Duplicated Data
    2. Missing Data
    3. Corrupt Data
    4. Inaccurate Data
    
This kind of data is known as Dirty Data

# 2. Messy Data

##  When the Data has structural issue is it known as Messy or Untidy Data.

Tidy Data has the following properties:
    
    1. Each variable forms a column
    2. Each observation forms a row
    3. Each observation unit forms a table

if a data violates any one of above it is known as untidy data

# Types of Assessment:-
    
## There are 2 types of assessment styles
    
    1. Manual - Looking through the data manually in google sheets
    2. Programmatic - By using pandas function such as info(), describe(), or sample()

# Steps in Assessment

## There are 2 steps involved in Assessment

1. Discover
2. Document


## Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
advers = pd.read_csv('adverse_reactions.csv')

In [4]:
print(patients.shape)
print(treatments.shape)
print(advers.shape)

(503, 14)
(280, 7)
(34, 3)


In [5]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [6]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [7]:
advers.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [8]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

## Quality

[Completeness, Validity, Accuracy, Consistency]

**Patients***

- Zip code column is stored in decimal format [V]
- Certain Zip code are of 4 digit [V]
- For the state column sometimes full names are used, otherwise short form use [Consistency Issues]
- Missing values fr contact details of 12 patients [Completeness]
- Incorrect data types for assigned_sex and birthdate Column [?]
- Errorneous data for weights and heights [Accuracy]
- Inconsistents way of storing phone number in contact column [Consistency Issues]
- Duplicate data & not useful data in datasets [Consistency Issues]

**Treatments**
 
- hba1c_change column has missing Value [Completeness]
- Novodra and Auralin column has attached to number [V]
- Only 280 rows are there, instead of 350 [Completeness]
- Name and surname column is in lower case [Consistency Issues]
- Incorrect data type for auralin and novadra column [V]
- hba1c column change 9 instead of 4 [Accuracy]
- Use of dashes instead of Nan in aurolin and novodra columns [Consistency Issues]
 
**Adverse Reactions**

- Name and surname column is in lower case [Consistency Issues]



In [3]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [4]:
patients['address'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
498    False
499    False
500    False
501    False
502    False
Name: address, Length: 503, dtype: bool

In [5]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [9]:
advers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [10]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [11]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [12]:
patients.sample()

patient_id assigned_sex given_name surname               address  \
254         255         male       Kang     Mai  1109 Beechwood Drive   

           city state  zip_code        country  \
254  Pittsburgh    PA   15205.0  United States   

                                contact  birthdate  weight  height   bmi  
254  412-274-6756KangMai@jourrapide.com  9/16/1986   200.4      70  28.8

In [13]:
patients.sample(5)

patient_id assigned_sex given_name       surname            address  \
317         318       female      Nancy        Parker   4605 Hall Street   
165         166         male     Zlatko      Rukavina      592 Rafe Lane   
25           26         male     Gregor          Bole  922 Chapmans Lane   
141         142       female     Evelyn    Uspenskaya   944 Thrash Trail   
214         215       female   Gabriela  Michalčíková  383 Marion Street   

            city state  zip_code        country  \
317    Las Vegas    NV   89110.0  United States   
165   Yazoo City    MS   39194.0  United States   
25   Albuquerque    NM   87109.0  United States   
141       Dallas    TX   75207.0  United States   
214     Hartford    VT    5047.0  United States   

                                             contact   birthdate  weight  \
317          NancyJParker@gustr.com+1 (702) 438-5138   3/21/1945   106.0   
165         ZlatkoRukavina@cuvox.de+1 (662) 716-9586  12/12/1976   198.7   
25                  GregorBole@gustr.com505-828-4955   6/19/1922   180.8   
141           903-644-2098EvelynUspenskaya@rhyta.com  10/25/1934   110.2   
214  GabrielaMichalcikova@rhyta.com+1 (802) 952-6490   10/8/1964   126.9   

     height   bmi  
317      63  18.8  
165      70  28.5  
25       67  28.3  
141      61  20.8  
214      65  21.1

In [14]:
patients.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
498    False
499    False
500    False
501    False
502    False
Length: 503, dtype: bool

In [15]:
patients['address'].duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
498    False
499    False
500    False
501    False
502     True
Name: address, Length: 503, dtype: bool

In [17]:
# List of person who is duplicated

patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

## Tidiness

**Patients**

- Phone and email together in one column i.e.. contact

**Treatments**
- 3 Variables stored in auralin and novodra column 

**Adverse Reactions**

- No Point of using these tables


## Cleaning Data

You should follow the order provided below:-
   - Completeness Issues
   - Tidiness Issues
   - Remaining Data Quality Issues like validity, Accuracy and Consistency

# Steps Involve in Data Cleaning

- Define
- Code
- Test

**Always make sure to create a copy of your pandas dataframe before you start the cleaning proces**


In [22]:
#Creating copies

Patients_copy = patients.copy()
treatments_copy = treatments.copy()
advers_copy = advers.copy()

In [23]:
Patients_copy[Patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

###  a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change column has missing values



#### Define

In [24]:
# Subtract hba1c_start from hba1c_end to get hba1c_Change

#### Code

In [26]:
treatments_copy['hba1c_Change'] = treatments_copy['hba1c_start'] - treatments_copy['hba1c_end']

#### Test

In [27]:
treatments_copy.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    109
hba1c_Change      0
dtype: int64

In [28]:
treatments_copy.describe()

hba1c_start   hba1c_end  hba1c_change  hba1c_Change
count   280.000000  280.000000    171.000000    280.000000
mean      7.985929    7.589286      0.546023      0.396643
std       0.568638    0.569672      0.279555      0.060168
min       7.500000    7.010000      0.200000      0.200000
25%       7.660000    7.270000      0.340000      0.350000
50%       7.800000    7.420000      0.380000      0.390000
75%       7.970000    7.570000      0.920000      0.440000
max       9.950000    9.580000      0.990000      0.540000

###  b. Tideness Issue

- Phone + email in contact column patients
- Novodra and Aurlin column treatment
- Merge the adverse table to treatment table



#### Define

In [31]:
# Novodra and Aurolin using the melt function

#### Code

In [34]:
treatments_copy = pd.melt(treatments_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
       var_name='treatment', value_name='dose')

In [36]:
treatments_copy

given_name       surname  hba1c_start  hba1c_end  hba1c_change  \
0     veronika      jindrová         7.63       7.20           NaN   
1       elliot    richardson         7.56       7.09          0.97   
2     yukitaka      takenaka         7.68       7.25           NaN   
3         skye    gormanston         7.97       7.62          0.35   
4       alissa        montez         7.78       7.46          0.32   
..         ...           ...          ...        ...           ...   
835     albina      zetticci         7.93       7.73          0.20   
836       john   teichelmann         7.90       7.58           NaN   
837     mathea       lillebø         9.04       8.67          0.37   
838     vallie        prince         7.64       7.28          0.36   
839     samúel  guðbrandsson         8.00       7.64          0.36   

        treatment       dose  
0         auralin  41u - 48u  
1         auralin          -  
2         auralin          -  
3         auralin  33u - 36u  
4         auralin          -  
..            ...        ...  
835  hba1c_Change        0.2  
836  hba1c_Change       0.32  
837  hba1c_Change       0.37  
838  hba1c_Change       0.36  
839  hba1c_Change       0.36  

[840 rows x 7 columns]

In [37]:
treatments_copy = treatments_copy[treatments_copy['dose']!='-']

In [39]:
treatments_copy['start'],treatments_copy['end'] = treatments_copy['dose'].str.split('-').str

C:\Users\hd\AppData\Local\Temp/ipykernel_3172/747519088.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_copy['start'],treatments_copy['end'] = treatments_copy['dose'].str.split('-').str
C:\Users\hd\AppData\Local\Temp/ipykernel_3172/747519088.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_copy['start'],treatments_copy['end'] = treatments_copy['dose'].str.split('-').str


In [40]:
treatments_copy

given_name       surname  hba1c_start  hba1c_end  hba1c_change  \
0     veronika      jindrová         7.63       7.20           NaN   
3         skye    gormanston         7.97       7.62          0.35   
6       sophia        haugen         7.65       7.27          0.38   
7        eddie        archer         7.89       7.55          0.34   
9         asia       woźniak         7.76       7.37           NaN   
..         ...           ...          ...        ...           ...   
835     albina      zetticci         7.93       7.73          0.20   
836       john   teichelmann         7.90       7.58           NaN   
837     mathea       lillebø         9.04       8.67          0.37   
838     vallie        prince         7.64       7.28          0.36   
839     samúel  guðbrandsson         8.00       7.64          0.36   

        treatment       dose start   end  
0         auralin  41u - 48u  41u    48u  
3         auralin  33u - 36u  33u    36u  
6         auralin  37u - 42u  37u    42u  
7         auralin  31u - 38u  31u    38u  
9         auralin  30u - 36u  30u    36u  
..            ...        ...   ...   ...  
835  hba1c_Change        0.2   NaN   NaN  
836  hba1c_Change       0.32   NaN   NaN  
837  hba1c_Change       0.37   NaN   NaN  
838  hba1c_Change       0.36   NaN   NaN  
839  hba1c_Change       0.36   NaN   NaN  

[560 rows x 9 columns]

In [42]:
treatments_copy.drop(columns = {'dose'}, inplace= True)

C:\Users\hd\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


####  Test

In [44]:
treatments_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20           NaN   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37           NaN   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u